keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,989 @@
+
+4. **Indefinite rerouting of the request**
+    
+    We should ensure at most one proxy, rather than proxying the request over and over again (if the source apiserver has an incorrect understanding of what the destination apiserver can serve). To do this, we will add a new header such as `X-Kubernetes-UVIP-Rerouted:true` to the  request once it is determined that the request cannot be served by the local apiserver and should therefore be proxied.  
+    We will remove this header after the request is received by the destination apiserver (i.e. after the proxy has happened once) at which point it will be served locally.
+
+5. **Putting IP/endpoint and trust bundle control in user hands in REST APIs**
+    
+    To prevent server-side request forgeries we will not give control over information about apiserver IP/endpoint and the trust bundle (used to authenticate server while proxying) to users via REST APIs.
+
+## Design Details
+
+### Aggregation Layer
+
+![Alt text](https://user-images.githubusercontent.com/26771552/243512491-47c79882-1f32-4067-8e10-f8e16faa4997.png?raw=true "Optional Title")
+
+1. A new filter will be added to the [handler chain] of the aggregation layer. This filter will maintain an internal map with the key being the group-version-resource and the value being a list of server IDs of apiservers that are capable of serving that group-version-resource
+   1. This internal map is populated using an informer for StorageVersion objects. An event handler will be added for this informer that will get the apiserver ID of the requested group-version-resource and update the internal map accordingly
+
+2. This filter will pass on the request to the next handler in the local aggregator chain, if:
+   1. It is a non resource request
+   2. The StorageVersion informer cache hasn't synced yet or if `StorageVersionManager.Completed()` has returned false. We will serve error 503 in this case
+   3. The request has a header `X-Kubernetes-UVIP-Rerouted:true` that indicates that this request has been proxied once already. If for some reason the resource is not found locally, we will serve error 503
+   4. No StorageVersion was retrieved for it, meaning the request is for an aggregated API or for a custom resource
+   5. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 404 with error `GVR <group_version_resource> is not served by anything in this cluster`
```

Comment:
This sounds racey.  What if storage version objects have not been created yet?

Reply:
> Ah I have missed to include a section about having poststarthook that will ensure that the storage version informer has synced before we mark the apiserver as ready

_And_ that we do not serve _any_ requests until we are done creating/updating SV objects.

Reply:
Added.